File: gradings/labs/lab1/valeev-aidar/lang_m/src/src/Parcer.cs
using System;

namespace src
{
    public class Parcer
    {
        public string ParcerResult;

        private string _input;
        private Token _lookahead;
        private int _yytext;
        private int _yyleng;

        public enum Token
        {
            INIT,
            EOI,
            SEMI,
            PLUS,
            TIMES,
            LP,
            RP,
            NUM,
            ERR
        }

        public Parcer(string input)
        {
            ParcerResult = "";
            _input = input;
            _lookahead = Token.INIT;
        }

        public Token Lex()
        {
            var current = _yytext + _yyleng;

            while (true)
            {
                if (current == _input.Length)
                    return Token.EOI;

                while (_input[current] == ' ' || _input[current] == '\n')
                {
                    current++;

                    if (current == _input.Length)
                        return Token.EOI;
                }

                for (; current < _input.Length; current++)
                {
                    _yytext = current;
                    _yyleng = 1;

                    switch (_input[current])
                    {
                        case ';': return Token.SEMI;
                        case '+': return Token.PLUS;
                        case '*': return Token.TIMES;
                        case '(': return Token.LP;
                        case ')': return Token.RP;

                        case '\r':
                        case '\n':
                        case '\t':
                        case ' ': break;

                        default:
                            if (char.IsDigit(_input[current]))
                            {
                                while (current < _input.Length && char.IsDigit(_input[current]))
                                    current++;

                                _yyleng = current - _yytext;

                                return Token.NUM;
                            }
                            else
                            {
                                Console.WriteLine("a");
                                Console.WriteLine("z" + _input[current] + "Z");
                                Console.WriteLine("b");
                                Console.ReadKey();
                                return Token.ERR;
                            }
                    }
                }
            }
        }

        private bool Match(Token token)
        {
            if (_lookahead == Token.INIT)
                Advance();

            return token == _lookahead;
        }

        private void Advance()
        {
            _lookahead = Lex();
        }

        private void Expression()
        {
            Term();
            ExprPrime();
        }

        private void Term()
        {
            Factor();
            TermPrime();
        }

        private void ExprPrime()
        {
            if (Match(Token.PLUS))
            {
                ParcerResult += "PLUS ";
                Advance();
                Term();
                ExprPrime();
            }
        }

        private void TermPrime()
        {
            if (Match(Token.TIMES))
            {
                ParcerResult += "TIMES ";
                Advance();
                Factor();
                TermPrime();
            }
        }

        private void Factor()
        {
            if (!Match(Token.LP))
            {
                ParcerResult += "NUM ";
                Advance();
            }
            else
            {
                ParcerResult += "LP ";
                Advance();
                Expression();
                ParcerResult += "RP ";
                Advance();
            }
        }

        public void Statements()
        {
            Expression();

            if (Match(Token.SEMI))
            {
                ParcerResult += "SEMI ";
                Advance();
            }

            if (!Match(Token.EOI))
            {
                Statements(); /* Do another statement. */
            }
            else
            {
                ParcerResult += "EOF";
            }
        }
    }
}

File: labs/lab1/src/parcer.c
//
// Created by cubazis on 27.08.18.
//

#include "parcer.h"
#include "string.h"

/** Variables for inter communication of lexical analyzer and parser */

char *yytext  = "";   /* Lexeme (not '\0' terminated)   */
int yyleng    = 0;    /* Lexeme length.                 */
int yylineno  = 0;    /* Input line number              */

void* lex()
{
	static char buffer[128];
	char        *current;

	current = yytext + yyleng;  	/* Skip current lexeme  */

	while( 1 )                  	/* Get the next one     */
	{
		while( !*current )
		{
			/* Get new lines, skipping any leading white space on the line,
			 * until a nonblank line is found.
			 */

			current = buffer;
			if( !fgets( buffer, sizeof(buffer), file) )
			{
				*current = '\0' ;
				return _EOI;
			}
			++yylineno;

			while( isspace(*current) )
				++current;
		}

		for( ; *current ; ++current )
		{
			/* Get the next token */

			yytext = current;
			yyleng = 1;

			switch( *current )
			{
				case ';': return (void *) _SEMI;
				case '+': return (void *)_PLUS  ;
				case '*': return (void *)_TIMES ;
				case '(': return (void *)_LP    ;
				case ')': return (void *)_RP    ;

				case '\n':
				case '\t':
				case '\r':
				case ' ' : break;

				default:
					if( !isdigit(*current) ){
						//fprintf(stderr, "Ignoring illegal input <%c>\n", *current);
						return (void *)_ERR_1;
					}
					else
					{
						while( isdigit(*current) )
							++current;

						yyleng = (int) (current - yytext);
						return (void *)_NUM;
					}
			}
		}
	}
}

static int Lookahead = -1;      /* Lookahead token  */

int match(int token )
{
	/* Return true if "token" matches the current lookahead symbol. */

	if( Lookahead == -1 )
		Lookahead = (int) lex();

	return token == Lookahead;
}

void advance()
{
	/* Advance the lookahead to the next input symbol.  */

	Lookahead = (int) lex();
}

void expression()
{
	/* expression -> term expression' */

	/** YOUR CODE HERE */
}

void term()
{
	/* term -> factor term' */

	/** YOUR CODE HERE */
}

void expr_prime()
{
	/* expression' -> PLUS term expression'
	 *              | epsilon
	 */

	/** YOUR CODE HERE */
}

void term_prime()
{
	/* term' -> TIMES factor term'
	 *       |   epsilon
	 */

	/** YOUR CODE HERE */
}

void factor()
{
	/* factor   ->    NUM_OR_ID
	 *          |     LP expression RP
	 */
	/** YOUR CODE HERE */
}

void statements()
{
	/*  statements -> expression SEMI
	 *             |  expression SEMI statements
	 */
	expression();

	if( match( _SEMI ) ){
		strcat(parser_result, "SEMI ");
		advance();
	}

	if( !match(_EOI) ){
		statements();			/* Do another statement. */
	} else {
		strcat(parser_result, "EOF");
	}
}
